File: ruby/binary.rb
class Binary

  def self.to_decimal(binary_string, total=0)
    raise ArgumentError, "Invalid binary string" if binary_string.match(/[^01]/)
    binary_string.split("").reverse.each_with_index do |n, index|
      n=="0" ? next : total+=(2 ** index)
    end
    total
  end

end

# tried to use .map, but then I couldn't gauge the index of all the 1s...

File: js/binary.js
Binary = function(number){
  this.number = number;
};

Binary.prototype.toDecimal = function() {
  if (this.number.match(/[^01]/)){return 0};
  values = this.number.split("").reverse().map(function(number, index){
    return number==="1" ? Math.pow(2, index) : parseInt(number)
  })
  return values.reduce((x,y) => x+y);
};

module.exports = Binary;
